llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -290,6 +293,62 @@ bool AIE2PreLegalizerCombinerImpl::tryToCombineVectorInserts(
   return true;
 }
 
+/// \returns true if it is possible to combine the below sequence of MIRs
+/// From : %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem16.I512),
+/// %2(<32 x s16>), %0(s32), %1(s32)
+///        %20:_(s16) = G_TRUNC %10(s32)
+///        %30:_(s20) = G_SEXT %20(s16)
+/// To :   %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem16.I512),
+/// %2(<32 x s16>), %0(s32), %1(s32)
+///        %30:_(s20) = G_TRUNC %10(s32)
+/// Or even:
+/// From : %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem8.I512),
+/// %2(<64 x s8>), %0(s32), %1(s32)
+///        %20:_(s8) = G_TRUNC %10(s32)
+///        %30:_(s20) = G_SEXT %20(s8)
+/// To :   %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem8.I512),
+/// %2(<64 x s8>), %0(s32), %1(s32)
+///        %30:_(s20) = G_TRUNC %10(s32)
+/// This also enables S20Narrowing for vextract
+bool AIE2PreLegalizerCombinerImpl::tryToCombineTruncExt(
```

Comment:
Instead of doing this combine here, why don't we just use `G_ASSERT_ZEXT`/`G_ASSERT_SEXT`?
In your previous commit 16445fb7d1ae5ecec7584ac2780cbd3a1310da4c,
we should pre-select
```
%0:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem8.I512), ...
```
into
```
%new:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem8.I512)
%0:_(s32) = G_ASSERT_ZEXT %new, 16
```

Then you should get the G_TRUNC + G_ZEXT combine for free.

Reply:
We should have something like:
```
    %0:_(s32) = G_CONSTANT i32 7
    %2:_(<32 x s16>) = COPY $x0
    %9:_(s20) = G_AIE_ZEXT_EXTRACT_VECTOR_ELT %2(<32 x s16>), %0(s32)
    %3:_(s20) = G_ASSERT_ZEXT %9, 16
    %4:_(s16) = G_TRUNC %3(s20)
    %5:_(s20) = G_ZEXT %4(s16)
    %6:_(p0) = G_CONSTANT i20 0
    %7:_(p0), %8:_(s20) = G_INTRINSIC intrinsic(@llvm.aie2.add.2d), %6(p0), %5(s20), %5(s20), %5(s20), %5(s20)
    PseudoRET implicit $lr, implicit %7(p0)
```

However, for SEXT case, there is no combine pattern AFAIK.

Reply:
We can try to match this pattern explicitly looking to ASSERT, something like:

```
bool CombinerHelper::matchCombineSextTrunc(MachineInstr &MI, Register &Reg) {
  assert(MI.getOpcode() == TargetOpcode::G_SEXT && "Expected a G_SEXT");
  Register DstReg = MI.getOperand(0).getReg();
  Register SrcReg = MI.getOperand(1).getReg();
  LLT DstTy = MRI.getType(DstReg);
  if (mi_match(SrcReg, MRI,
               m_GTrunc(m_all_of(m_Reg(Reg), m_SpecificType(DstTy))))) {
    unsigned DstSize = DstTy.getScalarSizeInBits();
    unsigned SrcSize = MRI.getType(SrcReg).getScalarSizeInBits();
    MachineInstr* MIDef = MRI.getVRegDef(Reg);
    if (MIDef->getOpcode() != TargetOpcode::G_ASSERT_SEXT)
      return false;
    unsigned ExtBits = MIDef->getOperand(2).getImm();
    return SrcSize == ExtBits;
  }
  return false;
}
```

And apply `replaceSingleDefInstWithReg`

Reply:
Instead of hardcoding the G_ASSERT_SEXT, I'd suggest to use the KnownBits analysis to check if we have known sign bits: ` KB->computeNumSignBits(Reg)`

Reply:
Above pattern is not matching in our case as m_SpecificType is s20 and DstTy is s32.
Even for ZEXT is not matching here, but combining is getting done somewhere later to pre combiner

Reply:
Hi @abhinay-anubola , all you need to do is to translate the intrinsic to:

```
    %9:_(s20) = G_AIE_ZEXT_EXTRACT_VECTOR_ELT %2(<32 x s16>), %0(s32)
    %3:_(s20) = G_ASSERT_ZEXT %9, 16
    %10:_(s32) = G_ZEXT %3(s20)
```

Globally speaking you should have:

```
    %0:_(s32) = G_CONSTANT i32 7
    %2:_(<32 x s16>) = COPY $x0
    %9:_(s20) = G_AIE_ZEXT_EXTRACT_VECTOR_ELT %2(<32 x s16>), %0(s32)
    %3:_(s20) = G_ASSERT_ZEXT %9, 16
    %10:_(s32) = G_ZEXT %3(s20)
    %4:_(s16) = G_TRUNC %10(s32)
    %5:_(s20) = G_ZEXT %4(s16)
    %6:_(p0) = G_CONSTANT i20 0
    %7:_(p0), %8:_(s20) = G_INTRINSIC intrinsic(@llvm.aie2.add.2d), %6(p0), %5(s20), %5(s20), %5(s20), %5(s20)
    PseudoRET implicit $lr, implicit %7(p0)
```

That can be automatically combined to:

```
    %0:_(s32) = G_CONSTANT i32 7
    %2:_(<32 x s16>) = COPY $x0
    %9:_(s20) = G_AIE_ZEXT_EXTRACT_VECTOR_ELT %2(<32 x s16>), %0(s32)
    %3:_(s20) = G_ASSERT_ZEXT %9, 16
    %6:_(p0) = G_CONSTANT i20 0
    %7:_(p0), %8:_(s20) = G_INTRINSIC intrinsic(@llvm.aie2.add.2d), %6(p0), %3(s20), %3(s20), %3(s20), %3(s20)
    PseudoRET implicit $lr, implicit %7(p0)
```

For the SEXT case, we need a new combiner looking to  `computeNumSignBits`.

Reply:
Hi @abhinay-anubola, just have in mind that we need to filter the cases: just consider cases with single consumer (s20).